### PR TITLE
Add LRO information to Mirrors

### PR DESCRIPTION
**Pull request by gesquivelgaghi**
## Summary

Summarize the changes made in this pull request. Include any relevant context or background information that would help reviewers understand the purpose and scope of the changes.

## Release Impact

Pick the target branch (see [RELEASES.md](../RELEASES.md) for the full model):

- [ ] `dev` — internal testing → publishes to `ppa-build-dev`
- [x] `main` — next beta → publishes to `ppa-build`
- [ ] `release/YY.MM` — point release on a maintained cycle → publishes to `ppa-build-YY.MM` (and to `ppa-build-stable` if this is the currently-promoted branch). Specify cycle: `release/____`

Change type (tick one):

- [ ] Patch (fix)
- [x] Minor (feature)
- [ ] Major (breaking)

> The change-type label is informational and only affects how the entry is rendered in the CHANGELOG. The actual version is computed from the branch and CalVer cycle — `pnpm changeset`'s `patch`/`minor` choice does not influence it.

## Checklist

- [x] **Changeset added** — I have run `pnpm changeset` (or `pnpm changeset --empty` if no CHANGELOG line is warranted) and committed the resulting `.md` file. Required for PRs targeting `main` and `release/*`; enforced by the `Changeset check` workflow.
- [x] **UI verified** — I have verified the changes locally.
- [x] **Linting clean** — No linting errors are present (especially in `scripts/`).

## Versioning Reminder

> [!IMPORTANT]
> Landscape UI uses **CalVer** (`YY.MM.Point.Build`). Version derivation by branch:
>
> - `main` / `point/*` → `YY.MM.0.<run>-beta`
> - `dev` → `YY.MM.0.<run>-dev`
> - `release/YY.MM` → `YY.MM.1.<run>` (cycle pinned by branch name)

### CODE REVIEW

File: src/features/operations/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  if (status === "succeeded") {
+    return "success-grey";
+  }
+
+  if (status === "failed") {
+    return ICONS.error;
+  }
+
+  return `${ICONS.spinner} u-animation--spin`;
```
[the remark this answers is not in the document]
Comment (gesquivelgaghi) [reply src/features/operations/helpers.ts:19]: the name prop translates to a className so this works fine

File: src/features/operations/components/ViewLogsSidePanel/ViewLogsSidePanel.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
+import SidePanel from "@/components/layout/SidePanel";
+import usePageParams from "@/hooks/usePageParams";
+import {
+  Button,
+  CodeSnippet,
+  Icon,
+  Notification,
+} from "@canonical/react-components";
+import { useEffect, useRef, useState, type FC } from "react";
```
[the remark this answers is not in the document]
Comment (gesquivelgaghi) [reply src/features/operations/components/ViewLogsSidePanel/ViewLogsSidePanel.tsx]: it doesnt currently use it

File: src/features/mirrors/components/MirrorDetails/components/MirrorDetailsActionBlock/MirrorDetailsActionBlock.tsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
+import { useEffect, type FC } from "react";
+import { Button, Icon, ICONS, Tooltip } from "@canonical/react-components";
+import { useListPublicationTargets } from "../../../../api";
```
[the remark this answers is not in the document]
Comment (gesquivelgaghi) [reply src/features/mirrors/components/MirrorDetails/components/MirrorDetailsActionBlock/MirrorDetailsActionBlock.tsx:3]: the `useListPublications` usage was deleted, along with its import.

File: src/features/mirrors/components/MirrorsList/MirrorsList.tsx [hunk 4]
```diff
@@ -39,19 +62,38 @@ const MirrorsList: FC<MirrorsListProps> = ({ mirrors, emptyMsg }) => {
...
+              isGettingOperation={isGettingOperations}
+              operation={operation}
+            />
+          );
+        },
+        getCellIcon: ({ row: { original: mirror } }: CellProps<Mirror>) => {
+          // eslint-disable-next-line react/prop-types
+          const operation = operations[mirror.lastOperation ?? ""];
+          return getOperationStatusIcon(operation);
+        },
```
[the remark this answers is not in the document]
Comment (gesquivelgaghi) [reply src/features/mirrors/components/MirrorsList/MirrorsList.tsx:80]: it is addressing a weird error where eslint doesnt recognize the `lastOperation` field in `Mirror` even though it's in the type

File: src/features/mirrors/components/MirrorDetails/components/MirrorDetailsActionBlock/MirrorDetailsActionBlock.tsx [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+  const tryPublish = () => {
+    if (publicationTargets.length) {
+      setPageParams({
+        sidePath: [...sidePath, "publish"],
+      });
+    } else {
+      openNoPublicationTargetsModal();
+    }
+  };
```
[the remark this answers is not in the document]
Comment (gesquivelgaghi) [reply src/features/mirrors/components/MirrorDetails/components/MirrorDetailsActionBlock/MirrorDetailsActionBlock.tsx:51]: it does not and should not support publishing any publication without targets. If it did, it was a bug

File: src/features/operations/components/ViewLogsSidePanel/ViewLogsSidePanel.test.tsx [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+  <Suspense fallback={<LoadingState />}>
+    <ViewLogsSidePanel />
+    {isTestingCopy && <input type="textbox" />}
+  </Suspense>
+);
+
+describe("ViewLogsSidePanel", () => {
+  beforeEach(() => {
+    vi.clearAllMocks();
+  });
```
[the remark this answers is not in the document]
Comment (gesquivelgaghi) [reply src/features/operations/components/ViewLogsSidePanel/ViewLogsSidePanel.test.tsx:34]: the test passes so this isn't an issue

File: src/features/mirrors/components/MirrorDetails/MirrorDetails.tsx [hunk 7]
```diff
@@ -87,61 +67,29 @@ const MirrorDetails: FC = () => {
...
-            hasIcon
-            className="p-segmented-control__button"
-            onClick={openRemoveModal}
-          >
-            <Icon name={`${ICONS.delete}--negative`} />
-            <span className="u-text--negative">Remove</span>
-          </Button>
-        </div>
+            Your last mirror update was not completed successfully.
+          </Notification>
```
Comment (ethanashaw) [inline src/features/mirrors/components/MirrorDetails/MirrorDetails.tsx:87]: I think the `inline` prop would be more consistent with other notifications like this, maybe check with Dani though.

File: src/features/operations/components/ViewLogsSidePanel/ViewLogsSidePanel.tsx [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  const url = useMemo(() => {
+    const blob = new Blob([logs], { type: "text/plain" });
+    return URL.createObjectURL(blob);
+  }, [logs]);
+
+  useEffect(() => {
+    return () => {
+      window.clearTimeout(copiedTimeoutRef.current);
+    };
+  }, []);
```
Comment (ethanashaw) [inline src/features/operations/components/ViewLogsSidePanel/ViewLogsSidePanel.tsx:37]: What does this do? Would there ever be a timeout when the component initially mounts?

Comment (gesquivelgaghi) [reply src/features/operations/components/ViewLogsSidePanel/ViewLogsSidePanel.tsx:37]: no this is to clear the timeout when the component unmounts

File: src/features/mirrors/components/MirrorDetails/MirrorDetails.tsx (hunk 7 above)
Comment (gesquivelgaghi) [reply src/features/mirrors/components/MirrorDetails/MirrorDetails.tsx:87]: this is following the designs here https://www.figma.com/design/tf2IhkMdpOnSMc0wrHKDwi/26.10-Deb-Archive-LROs?node-id=9-1232&m=dev

also the inline looks weird with an action

**Review by ethanashaw — APPROVED** [review]

File: src/features/mirrors/components/MirrorsList/MirrorsList.tsx [hunk 9]
```diff
@@ -20,15 +28,30 @@ interface MirrorsListProps {
...
 
+  const operationNames = mirrors
+    .filter((mirror) => mirror.lastOperation)
+    .map((mirror) => mirror.lastOperation ?? "");
+
+  const { operations: operations, isGettingOperations } = useBatchGetOperations(
+    operationNames,
+    {
+      refetchInterval: ({ state }) =>
+        Object.values(state.data ?? {}).some((operation) => !operation.done)
```
Comment (yurii-vasyliev) [inline src/features/mirrors/components/MirrorsList/MirrorsList.tsx:39]: Polling can stop while a mirror is still updating. This predicate only inspects operations *present* in `state.data`, which is keyed by `operation.name`. If the `batchGet` response ever omits a requested name, that op drops out of the map and is treated as "done", so `refetchInterval` returns `false` and polling halts while the op is still running. Could you confirm `batchGet` returns exactly one entry per requested name? If that's not guaranteed, base the predicate on the requested `operationNames` (an absent/!done name should keep polling) — or at least document the assumption here.

File: src/features/mirrors/components/MirrorsList/MirrorsList.test.tsx [hunk 10]
```diff
@@ -22,17 +26,30 @@ describe("MirrorsList", () => {
...
+
+  it("renders status cell for no operation", () => {
+    const mirror = {
+      ...mirrors[0],
+      lastOperation: undefined,
+    };
+
+    renderWithProviders(<MirrorsList mirrors={[mirror]} />);
+
+    expect(screen.getByText("Not yet updated")).toBeInTheDocument();
```
Comment (yurii-vasyliev) [inline src/features/mirrors/components/MirrorsList/MirrorsList.test.tsx:53]: Test-coverage gap on the core new behavior. These tests cover "Updated" (succeeded) and "Not yet updated" (no op), but the mocks also include in-progress (`pppp-gggg-ssss`) and failed (`ffff-llll-dddd`) operations whose rendering is never asserted — the progress bar + "Updating %", and "Update failed" + the **View logs** action. Worth adding a case each for a failed and an in-progress `lastOperation`. Same gap in `MirrorDetails.test.tsx`, which currently only asserts the "Status" label renders, not the status value/icon or the failure `Notification`.

File: src/tests/server/handlers/operations.ts [hunk 11]
```diff
@@ -7,10 +7,48 @@ import {
   overCountOperation,
   timeoutOperation,
   idleOperation,
+  operations,
 } from "@/tests/mocks/operations";
 import { http, delay, HttpResponse } from "msw";
 
+let progress = inProgressOperation.metadata.progressPercent;
```
Comment (yurii-vasyliev) [inline src/tests/server/handlers/operations.ts:14]: This module-level `progress` counter isn't reset between tests. It only self-resets when it reaches >=100, and `server.resetHandlers()` doesn't touch module state, so a test that unmounts mid-poll leaves residual progress — a later test polling `pppp-gggg-ssss` can then flip to "succeeded" after fewer polls than intended (latent flakiness / pass-for-the-wrong-reason). Suggest exporting a reset and calling it from a global `afterEach`, or deriving progress per-request instead of from shared mutable state.

**Review by yurii-vasyliev — COMMENTED** [review]
A few follow-up notes on top of the approval — non-blocking. Nice feature; the LRO wiring (polling, loading/empty/error states, progress) is clean. Three actionable items inline (one polling edge case, two test-coverage/determinism gaps); a handful of nits I'll leave out of the diff.

File: src/features/mirrors/components/MirrorsList/MirrorsList.tsx (hunk 9 above)
Comment (gesquivelgaghi) [reply src/features/mirrors/components/MirrorsList/MirrorsList.tsx:39]: The batchGet request returns the list of operations based on their name, so the only case it wouldn't be included would be if the operation is not longer listed in the names, or if there's a backend bug that they'd need to fix. Do you think it's fine to keep as is? I can add a comment here if you think it's unclear

File: src/features/mirrors/components/MirrorsList/MirrorsList.test.tsx (hunk 10 above)
Comment (gesquivelgaghi) [reply src/features/mirrors/components/MirrorsList/MirrorsList.test.tsx:53]: The rest of these cases are covered under `OperationStatusCell.test.tsx` which is called both by `MirrorsList` and `MirrorDetails`, so I thought we didn't need to cover those cases again in these parent components. For these, I only covered the branches they have (i.e., "has a `lastOperation`" vs "does not have a `lastOperation`"). Do you think we should add more tests anyway?

edit: I realized I was actually missing the no operation case in `MirrorDetails` so I just added it to my Publications LRO branch

File: src/tests/server/handlers/operations.ts (hunk 11 above)
Comment (gesquivelgaghi) [reply src/tests/server/handlers/operations.ts:14]: I can do the reset in the next PR but I'm curious about deriving the progress per-request. How would that work? Bc the progress would still need to persist across polling requests

File: src/features/mirrors/components/MirrorsList/MirrorsList.tsx (hunk 9 above)
Comment (yurii-vasyliev) [reply src/features/mirrors/components/MirrorsList/MirrorsList.tsx:39]: Yeah, that's definitely fine to keep as-is Thanks!

File: src/features/mirrors/components/MirrorsList/MirrorsList.test.tsx (hunk 10 above)
Comment (yurii-vasyliev) [reply src/features/mirrors/components/MirrorsList/MirrorsList.test.tsx:53]: Yes, agreed. The one thing that isn't a cell concern: the failure Notification banner + ViewLogsButton in MirrorDetails.tsx is rendered by the parent, not by OperationStatusCell, so it's currently uncovered. This is not a blocker for this PR of course, but nice to have.

File: src/tests/server/handlers/operations.ts (hunk 11 above)
Comment (yurii-vasyliev) [reply src/tests/server/handlers/operations.ts:14]: Deferring the reset to a follow-up is fine by me, thank you.

**Review by yurii-vasyliev — APPROVED** [review]
LGTM!

I'd like to discuss some design decisions with Dani, but I don't want to block this PR from merging. We can open a new PR in the future if needed.
